[LinuxConsumption] Support mounting Azure file share with Run-From-Pkg

balag0 opened the pull request:
Currently either Azure File share can be mounted or app can use Run-From-Package but not both simultaneously. This PR adds support for this scenario but only for PowerShell apps.
If AzureFiles are configured, it will be mounted at /home. If Run-From-Pkg is configured the contents will be made available at /home/site/wwwroot.

Note: It is possible for /home/site/wwwroot to point to a shared remote share. So copying contents to /home/site/wwwroot can be disruptive for other running instances. As a workaround /local/sitepackages will be used as a per-instance local storage where the zip will extracted before it gets mounted to /home/site/wwwwroot.

Fixes https://github.com/Azure/azure-functions-host/issues/7252

<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves #7252

### Pull request checklist

* [X ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [X ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ X] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ X] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

pragnagopa commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -289,15 +289,46 @@ private async Task ApplyContext(HostAssignmentContext assignmentContext)
 
-            if ((pkgContext.IsScmRunFromPackage() && await pkgContext.BlobExistsAsync(_logger)) ||
-                (!pkgContext.IsScmRunFromPackage() && !string.IsNullOrEmpty(pkgContext.Url) && pkgContext.Url != "1"))
+            if (assignmentContext.IsWorkerRuntimePowerShell)
```
make this more generic - SupportBindingMoutApi? In future this method can expanded for other languages and eventually remove the check

balag0 replied:
Yes. But thinking we can make it more generic when we are closer to making the actual changes.

balag0 replied:
Renaming to SupportsAzureFileShareMount

pragnagopa commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -289,15 +289,46 @@ private async Task ApplyContext(HostAssignmentContext assignmentContext)
+                else
+                {
+                    _logger.LogWarning(
+                        $"No {nameof(EnvironmentSettingNames.AzureFilesConnectionString)} or {nameof(EnvironmentSettingNames.AzureFilesContentShare)} configured");
```
Move this log into `IsAzureFilesContentShareConfigured`

`logger.LogDebug(
                       $"{nameof(EnvironmentSettingNames.AzureFilesConnectionString)} IsNullOrEmpty: {string.IsNullOrEmpty(AzureFilesConnectionString)}. {nameof(EnvironmentSettingNames.AzureFilesContentShare)}: IsNullorEmpty{string.IsNullOrEmpty(AzureFilesContentShare)}");`

pragnagopa commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -289,15 +289,46 @@ private async Task ApplyContext(HostAssignmentContext assignmentContext)
+                }
+                else
+                {
+                    _logger.LogWarning(
```
For powershell - we should log error level as without the Azure Files share - ManagedDependencies will not be persisted

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/BashCommandHandler.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            using (_metricsLogger.LatencyEvent(MetricEventNames.LinuxContainerSpecializationZipExtract))
+            {
+                _logger.LogInformation($"Extracting files to '{scriptPath}'");
```
change LogInformation to LogDebug. Across the host all logs are system logs - we should use debug level. Debug level is default for Kusto. This will cause less noise in user logs in AppInsights

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/BashCommandHandler.cs:
```diff
@@ -0,0 +1,80 @@
+                    process.Start();
+                    var output = process.StandardOutput.ReadToEnd().Trim();
+                    var error = process.StandardError.ReadToEnd().Trim();
+                    process.WaitForExit();
```
We should pass in a timeout and check if process exited: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.waitforexit?view=netcore-3.1

Also see https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/ProcessManagement/WorkerProcess.cs#L194-L213 for an example

balag0 replied:
This PR is just moving the code as is to a new file. But opened a task to track this separately.
https://github.com/Azure/azure-functions-host/issues/7287

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/IBashCommandHandler.cs:
```diff
@@ -0,0 +1,12 @@
+    public interface IBashCommandHandler
+    {
+        void UnzipPackage(string filePath, string scriptPath);
+
```
`UnzipPackage` package does not belong here?

balag0 replied:
Agree. moving to new interface.

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/IRunFromPackageHandler.cs:
```diff
@@ -0,0 +1,18 @@
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Management.LinuxSpecialization
+{
+    public interface IRunFromPackageHandler
+    {
```
Move `UnzipPackage` to this interface?

balag0 replied:
Was originally here. but moving out to make testing possible. Ideally would have been in a new interface , but just put in IBashCommandHandler since it seemed related enough :)
But can create a new interface if you think that is better.

balag0 replied:
Moving to new interface.

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageCloudBlockBlobService.cs:
```diff
@@ -0,0 +1,43 @@
+            return await BlobExistsAsync(url, environmentVariableName, logger);
+        }
+
+        private static async Task<bool> BlobExistsAsync(string url, string environmentVariableName, ILogger logger)
```
This looks like it belongs to a static helper. If the idea here is to provide an abstraction, declare an interface.

balag0 replied:
yes thats the idea, but seemed like an interface was not necessary since we can achieve this using virtual method. So leaning towards leaving this as is.

pragnagopa replied:
We should use `virtual` methods in the context of base class , derived class. In this case, declaring interface is the right approach.

balag0 replied:
I think both are valid approaches and following Mathew's feedback from an earlier PR since I already have it implemented this way  :)

https://github.com/Azure/azure-functions-host/pull/2962#discussion_r194864519
Note that you can mock using virtuals, without defining interfaces. I.e. the IHostAssignmentContextReader/ILinuxContainerInitializationService/etc. interfaces aren't really necessary. Interfaces are useful if you expect to have different implementations in the future, but that isn't the case for any of these.

pragnagopa replied:
thanks for explanation.

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageHandler.cs:
```diff
@@ -0,0 +1,280 @@
+
+            var uri = new Uri(pkgContext.Url);
+            // check file name since it'll be faster than running `file`
+            if (FileIsAny(".squashfs", ".sfs", ".sqsh", ".img", ".fs"))
```
move the list of string to a static string[]

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageHandler.cs:
```diff
@@ -0,0 +1,280 @@
+            // Check file magic-number using `file` command.
+            (var output, _, _) = _bashCommandHandler.RunBashCommand($"{BashCommandHandler.FileCommand} -b {filePath}", MetricEventNames.LinuxContainerSpecializationFileCommand);
+            if (output.StartsWith("Squashfs", StringComparison.OrdinalIgnoreCase))
+            {
```
move strings to constants

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageHandler.cs:
```diff
@@ -0,0 +1,280 @@
+            var zipUri = new Uri(pkgContext.Url);
+            if (!Utility.TryCleanUrl(zipUri.AbsoluteUri, out string cleanedUrl))
+            {
+                throw new Exception("Invalid url for the package");
```
InvalidOperationException ?

balag0 replied:
Dont have a preference :) switching to InvalidOperationException

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageHandler.cs:
```diff
@@ -0,0 +1,280 @@
+            var filePath = Path.Combine(tmpPath, fileName);
+            if (pkgContext.PackageContentLength != null && pkgContext.PackageContentLength > AriaDownloadThreshold)
+            {
+                _logger.LogInformation($"Downloading zip contents from '{cleanedUrl}' using aria2c'");
```
nit: Use Log.Debug

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/RunFromPackageHandler.cs:
```diff
@@ -0,0 +1,280 @@
+            var filePath = Path.Combine(tmpPath, fileName);
+            if (pkgContext.PackageContentLength != null && pkgContext.PackageContentLength > AriaDownloadThreshold)
+            {
+                _logger.LogInformation($"Downloading zip contents from '{cleanedUrl}' using aria2c'");
```
As an improvement in future PRs - extract to a base class  and provide concrete implementation for each type of download we support - right now its Aria and HttpClient - this will make it easy to test and extend to other types download if needed.

balag0 replied:
Sure.

pragnagopa commented on src/WebJobs.Script.WebHost/Models/HostAssignmentContext.cs:
```diff
@@ -55,6 +56,21 @@ public string AzureFilesContentShare
                 ? Environment[EnvironmentSettingNames.AzureFilesContentShare]
                 : SiteName;
 
+        public string FunctionsWorkerRuntime =>
```
Move this to EnvironmentExtensions - so can be reused at other places as well.

balag0 replied:
Done.

pragnagopa commented on src/WebJobs.Script.WebHost/Models/RunFromPackageContext.cs:
```diff
@@ -32,29 +35,10 @@ public bool IsScmRunFromPackage()
-            }
-
-            return exists;
+            return (IsScmRunFromPackage() && await _runFromPackageCloudBlockBlobService.BlobExists(Url, EnvironmentVariableName, logger)) ||
```
Apologies for piling on this PR - with this check in place - can we address issue @pgombar was running into with scm run from package  for isReadOnlyFileSystem check?

balag0 replied:
Yes, We could. But I think better/easier to make that change as a separate PR.

pragnagopa commented on src/WebJobs.Script/Environment/EnvironmentSettingNames.cs:
```diff
@@ -27,6 +27,7 @@ public static class EnvironmentSettingNames
         public const string AppInsightsConnectionString = "APPLICATIONINSIGHTS_CONNECTION_STRING";
         public const string AppInsightsQuickPulseAuthApiKey = "APPINSIGHTS_QUICKPULSEAUTHAPIKEY";
         public const string FunctionsExtensionVersion = "FUNCTIONS_EXTENSION_VERSION";
+        public const string FunctionWorkerRuntime = "FUNCTIONS_WORKER_RUNTIME";
```
This is already defined here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Workers/Rpc/RpcWorkerConstants.cs#L10

Can you see if you use it?

balag0 replied:
Looks like this is where all the appsettings are defined. So adding it here. We can remove the one in Rpc constants and reference this file if needed.

pragnagopa commented on test/WebJobs.Script.Tests.Integration/Management/HostAssignmentContextTests.cs:
```diff
@@ -53,5 +54,66 @@ public void Returns_BYOS_EnvironmentVariables()
+
+        [Theory]
+        [InlineData(RpcWorkerConstants.PowerShellLanguageWorkerName, true)]
+        [InlineData(RpcWorkerConstants.DotNetLanguageWorkerName, false)]
```
add tests using strings with different cases

pragnagopa reviewed (commented):
Added few comments. Thanks.

pragnagopa commented:
> If AzureFiles are configured, it will be mounted at /home. If Run-From-Pkg is configured the contents will be made available at /home/site/wwwroot.

What happens if AzureFiles has a folder ` /home/site/wwwroot.` ? which one wins for function app content?

balag0 commented:
> > If AzureFiles are configured, it will be mounted at /home. If Run-From-Pkg is configured the contents will be made available at /home/site/wwwroot.
> 
> What happens if AzureFiles has a folder ` /home/site/wwwroot.` ? which one wins for function app content?

RunFromPackage takes precedence. If RunFromPackage is defined it will hide the contents in file share.

pragnagopa commented on src/WebJobs.Script.WebHost/Management/LinuxSpecialization/IZipHandler.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Management.LinuxSpecialization
+{
+    public interface IZipHandler
```
nit: rename to `IExtractZipHandler` ? or `IUnZipHandler

pragnagopa commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -459,5 +460,19 @@ public static string GetStorageSuffix(this IEnvironment environment)
             }
         }
+
+        public static string FunctionsWorkerRuntime(this IEnvironment environment)
```
rename to `GetFunctionsWorkerRuntime`

pragnagopa commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -289,15 +289,46 @@ private async Task ApplyContext(HostAssignmentContext assignmentContext)
+                else
+                {
+                    _logger.LogError(
+                        $"No {nameof(EnvironmentSettingNames.AzureFilesConnectionString)} or {nameof(EnvironmentSettingNames.AzureFilesContentShare)} configured. Azure FileShare will not be mounted");
```
Include : `For Powershell Functions, Managed Dependencies will not persisted across functions host instances.`

pragnagopa commented on src/WebJobs.Script.WebHost/Management/InstanceManager.cs:
```diff
@@ -289,15 +289,46 @@ private async Task ApplyContext(HostAssignmentContext assignmentContext)
+                    if (!blobContextApplied && azureFilesMounted)
+                    {
+                        _logger.LogWarning($"Failed to {nameof(_runFromPackageHandler.ApplyBlobPackageContext)}. Attempting to use local disk instead");
+                        await _runFromPackageHandler.ApplyBlobPackageContext(pkgContext, options.ScriptPath, false);
```
Info: RunFromPackage takes precedence. If RunFromPackage is defined it will hide the contents in file share needs to logged as warning - to avoid confusion

pragnagopa commented on src/WebJobs.Script.WebHost/Management/MeshServiceClient.cs:
```diff
@@ -98,6 +105,20 @@ public async Task NotifyHealthEvent(ContainerHealthEventType healthEventType, Ty
         }
 
+        public async Task CreateBindMount(string sourcePath, string targetPath)
+        {
```
This is a generic question : As it's a new new API on mesh process - how do you co-ordinate host release + mesh image release

balag0 replied:
Dont have a specific process. Just need to make sure the code can handle failures gracefully. Usually API PRs get merged first anyway if not the PR owner can follow up at release time, but I havent had to do this so far as things in general move slower in host repo.

pragnagopa reviewed (commented):
Added minor comments. Thanks!

pragnagopa commented:
@pgombar - Please hold off on your change to check blob exists - as this PR refactors same code path

pragnagopa reviewed (dismissed):
LGTM.

Tagging @divyagandhisethi  @yojagad  @ahmelsayed  for a final pass

balag0 commented:
@divyagandhisethi @ahmelsayed  Could you take a look from KubernetesPodController pov as well since the same method is used there as well.

balag0 committed:
[LinuxConsumption] Support mounting Azure file share with Run-From-Package

   Currently either Azure File share can be mounted or app can use Run-From-Package but not both simultaneously. This PR adds support for this scenario but only for PowerShell apps. If AzureFiles are configured, it will be mounted at /home. If Run-From-Pkg is configured the contents will be made available at /home/site/wwwroot.
   Note: It is possible for /home/site/wwwroot to point to a shared remote share. So copying contents to /home/site/wwwroot can be disruptive for other running instances. As a workaround /local/sitepackages will be used as a per-instance local storage where the zip will extracted before it gets mounted to /home/site/wwwwroot.

   Fixes https://github.com/Azure/azure-functions-host/issues/7252
